From 7fc3cb83a4be698f0ea0fdf0f28ffd04bb592b56 Mon Sep 17 00:00:00 2001
From: Danny Baumann <dannybaumann@web.de>
Date: Wed, 24 May 2017 12:28:57 +0200
Subject: [PATCH 17/20] Properly close body InputStreams.

Change Ibe41e1896d6867cae0ee080f1556d7fa9e88966d allowed keeping them
open temporarily for later usage. Later they were attempted to be closed,
but the reference obtained for closing them was fetched by calling
getInputStream() again, which created a new stream. Collect the streams
in a passed ArrayList instead so the correct references can be closed.

Change-Id: I2b581484776c17beb5dc0cc8661185942507ab55
---
 src/com/android/emailcommon/internet/MimeUtility.java    | 8 ++++++--
 .../android/emailcommon/utility/ConversionUtilities.java | 9 +++++----
 2 files changed, 11 insertions(+), 6 deletions(-)

diff --git a/src/com/android/emailcommon/internet/MimeUtility.java b/src/com/android/emailcommon/internet/MimeUtility.java
index 8ea29dd70..93765ff62 100644
--- a/src/com/android/emailcommon/internet/MimeUtility.java
+++ b/src/com/android/emailcommon/internet/MimeUtility.java
@@ -219,10 +219,12 @@ public class MimeUtility {
      * Reads the Part's body and returns a String based on any charset conversion that needed
      * to be done.
      * @param part The part containing a body
+     * @param outInputStreams A list of input streams the opened body stream should be added to.
+     *                        If null is passed the stream should be closed.
      * @return a String containing the converted text in the body, or null if there was no text
      * or an error during conversion.
      */
-    public static String getTextFromPart(Part part, boolean closeInput) {
+    public static String getTextFromPart(Part part, ArrayList<InputStream> outInputStreams) {
         InputStream in = null;
         ByteArrayOutputStream out = null;
         try {
@@ -276,7 +278,9 @@ public class MimeUtility {
             Log.e(LOG_TAG, "Unable to getTextFromPart " + e.toString());
         } finally {
             IOUtils.closeQuietly(out);
-            if (closeInput) {
+            if (outInputStreams != null && in != null) {
+                outInputStreams.add(in);
+            } else {
                 IOUtils.closeQuietly(in);
             }
         }
diff --git a/src/com/android/emailcommon/utility/ConversionUtilities.java b/src/com/android/emailcommon/utility/ConversionUtilities.java
index 790255399..0c8cbfdfa 100644
--- a/src/com/android/emailcommon/utility/ConversionUtilities.java
+++ b/src/com/android/emailcommon/utility/ConversionUtilities.java
@@ -23,6 +23,7 @@ import com.android.emailcommon.mail.Part;
 
 import android.text.TextUtils;
 
+import java.io.InputStream;
 import java.util.ArrayList;
 
 public class ConversionUtilities {
@@ -63,20 +64,20 @@ public class ConversionUtilities {
      */
     public static BodyFieldData parseBodyFields(ArrayList<Part> viewables)
             throws MessagingException {
-        return  parseBodyFields(viewables, true);
+        return  parseBodyFields(viewables, null);
     }
 
     /**
      * Parse body text (plain and/or HTML) from MimeMessage to {@link BodyFieldData}.
      */
-    public static BodyFieldData parseBodyFields(ArrayList<Part> viewables, boolean closeInputs)
-    throws MessagingException {
+    public static BodyFieldData parseBodyFields(ArrayList<Part> viewables,
+            ArrayList<InputStream> outInputStreams) throws MessagingException {
         final BodyFieldData data = new BodyFieldData();
         StringBuffer sbHtml = null;
         StringBuffer sbText = null;
 
         for (Part viewable : viewables) {
-            String text = MimeUtility.getTextFromPart(viewable, closeInputs);
+            String text = MimeUtility.getTextFromPart(viewable, outInputStreams);
             // Deploy text as marked by the various tags
             boolean isHtml = "text/html".equalsIgnoreCase(viewable.getMimeType());
 
-- 
2.17.1

